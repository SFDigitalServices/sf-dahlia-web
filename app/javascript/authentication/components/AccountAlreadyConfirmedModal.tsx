import { AppearanceStyleType, Button, t } from "@bloom-housing/ui-components"
import React from "react"
import { Dialog } from "@bloom-housing/ui-seeds"

const AccountAlreadyConfirmedModal = ({
  isOpen,
  onClose,
}: {
  isOpen: boolean
  onClose: () => void
}) => {
  return (
    <Dialog isOpen={isOpen} onClose={onClose}>
      <Dialog.Header>{t("signIn.accountAlreadyConfirmed.title")}</Dialog.Header>
      <Dialog.Content>{t("signIn.accountAlreadyConfirmed.p1")}</Dialog.Content>
      <Dialog.Footer>
        <Button type="submit" styleType={AppearanceStyleType.primary} onClick={onClose}>
          {t("t.ok")}
        </Button>
      </Dialog.Footer>
    </Dialog>
  )
}

export { AccountAlreadyConfirmedModal as default, AccountAlreadyConfirmedModal }
