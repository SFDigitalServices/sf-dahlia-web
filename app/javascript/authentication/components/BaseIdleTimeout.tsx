import React, { FunctionComponent, useEffect, useState } from "react"

import {
  AlertTypes,
  AppearanceStyleType,
  Button,
  Modal,
  setSiteAlertMessage,
} from "@bloom-housing/ui-components"

const PROMPT_TIMEOUT = 60000
const events = ["mousemove", "keypress", "scroll"]

function useIdleTimeout(timeoutMs: number, onTimeout: () => void) {
  useEffect(() => {
    let timer: number = (setTimeout(onTimeout, timeoutMs) as unknown) as number
    const restartTimer = () => {
      if (timer) {
        clearTimeout(timer)
      }
      timer = (setTimeout(onTimeout, timeoutMs) as unknown) as number
    }

    // Listen for any activity events & reset the timer when they are found
    if (typeof document !== "undefined") {
      events.forEach((event) => document.addEventListener(event, restartTimer, false))
    }

    // Clean up our listeners & clear the timeout on unmounting/updating the effect
    return () => {
      if (timer) {
        clearTimeout(timer)
      }
      events.forEach((event) => document.removeEventListener(event, restartTimer, false))
    }
  }, [timeoutMs, onTimeout])
}

type IdleTimeoutProps = {
  promptTitle: string
  promptText: string
  promptAction: string
  redirectPath: string
  alertMessage: string
  alertType?: AlertTypes
  onTimeout: () => unknown
}

const BaseIdleTimeout: FunctionComponent<IdleTimeoutProps> = (props: IdleTimeoutProps) => {
  const {
    promptTitle,
    promptAction,
    promptText,
    redirectPath,
    alertMessage,
    alertType = "alert",
    onTimeout,
  } = props

  // 5 minutes
  const idleTimeout = 5 * 60 * 1000
  const [promptTimeout, setPromptTimeout] = useState<number | undefined>()

  useIdleTimeout(idleTimeout, () => {
    // Clear any existing prompt timeouts
    if (promptTimeout) {
      clearTimeout(promptTimeout)
    }

    const timeoutAction = async () => {
      setPromptTimeout(undefined)
      await onTimeout()
      setSiteAlertMessage(alertMessage, alertType)

      // replace this with proper react router when we have one
      return (window.location.href = redirectPath)
    }

    // Give the user 1 minute to respond to the prompt before the onTimeout action
    setPromptTimeout(
      (setTimeout(() => {
        void timeoutAction()
      }, PROMPT_TIMEOUT) as unknown) as number
    )
  })

  const modalActions = [
    <Button
      styleType={AppearanceStyleType.primary}
      onClick={() => {
        clearTimeout(promptTimeout)
        setPromptTimeout(undefined)
      }}
    >
      {promptAction}
    </Button>,
  ]

  return (
    <Modal
      open={Boolean(promptTimeout)}
      title={promptTitle}
      ariaDescription={promptText}
      actions={modalActions}
      hideCloseIcon
    >
      {promptText}
    </Modal>
  )
}

export default BaseIdleTimeout
