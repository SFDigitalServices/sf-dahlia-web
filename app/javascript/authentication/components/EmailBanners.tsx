import { t } from "@bloom-housing/ui-components"
import { Alert } from "@bloom-housing/ui-seeds"
import React from "react"
import { confirmEmail } from "../../api/authApiService"

export const requestEmail = (
  email: string,
  setEmailSent: (value: boolean) => void,
  setEmailSentError: (value: string | null) => void
) => {
  confirmEmail(email)
    .then(() => {
      setEmailSent(true)
    })
    .catch(() => {
      setEmailSentError(t("signIn.newAccount.sendEmailAgainButton.error"))
    })
}

export const EmailBanners = ({
  emailSent,
  emailSentError,
}: {
  emailSent: boolean
  emailSentError: string
}) => {
  return (
    <>
      {emailSent && (
        <Alert className="sign-in-banner banner-background-color">
          {t("signIn.newAccount.sendEmailAgainButton.confirmation")}
        </Alert>
      )}
      {emailSentError && (
        <Alert variant="alert" className="sign-in-banner">
          {emailSentError}
        </Alert>
      )}
    </>
  )
}
