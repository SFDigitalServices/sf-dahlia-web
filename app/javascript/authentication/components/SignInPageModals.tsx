import { AppearanceStyleType, Button, t } from "@bloom-housing/ui-components"
import React, { useState } from "react"
import { confirmEmail } from "../../api/authApiService"
import { Alert, Dialog } from "@bloom-housing/ui-seeds"

export const AccountAlreadyConfirmedModal = ({
  isOpen,
  onClose,
}: {
  isOpen: boolean
  onClose: () => void
}) => {
  return (
    <Dialog isOpen={isOpen} onClose={onClose}>
      <Dialog.Header>{t("signIn.accountAlreadyConfirmed.title")}</Dialog.Header>
      <Dialog.Content>{t("signIn.accountAlreadyConfirmed.p1")}</Dialog.Content>
      <Dialog.Footer>
        <Button type="submit" styleType={AppearanceStyleType.primary} onClick={onClose}>
          {t("t.ok")}
        </Button>
      </Dialog.Footer>
    </Dialog>
  )
}

export const NewAccountNotConfirmedModal = ({
  email,
  onClose,
}: {
  email: string
  onClose: () => void
}) => {
  const [emailSent, setEmailSent] = useState(false)
  const [emailSentError, setEmailSentError] = useState<string | null>(null)
  const requestEmail = () => {
    confirmEmail(email)
      .then(() => {
        setEmailSent(true)
      })
      .catch(() => {
        setEmailSentError(t("signIn.newAccount.sendEmailAgainButton.error"))
      })
  }

  return (
    <Dialog isOpen={!!email} onClose={onClose}>
      <Dialog.Header>{t("signIn.newAccount.title")}</Dialog.Header>
      {emailSent && (
        <Alert className="sign-in-banner banner-background-color">
          {t("signIn.newAccount.sendEmailAgainButton.confirmation")}
        </Alert>
      )}
      {emailSentError && (
        <Alert variant="alert" className="sign-in-banner">
          {emailSentError}
        </Alert>
      )}
      <Dialog.Content>{t("signIn.newAccount.p1", { email })}</Dialog.Content>
      <Dialog.Content>{t("signIn.newAccount.p2")}</Dialog.Content>
      <Dialog.Footer>
        <Button type="submit" styleType={AppearanceStyleType.primary} onClick={onClose}>
          {t("t.ok")}
        </Button>
        <Button styleType={AppearanceStyleType.secondary} onClick={requestEmail}>
          {t("signIn.newAccount.sendEmailAgainButton")}
        </Button>
      </Dialog.Footer>
    </Dialog>
  )
}
