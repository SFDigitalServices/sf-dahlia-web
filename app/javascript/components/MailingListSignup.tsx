import React, { useContext } from "react"
import { ActionBlock, Heading, Icon, t } from "@bloom-housing/ui-components"
import Link from "../navigation/Link"
import ConfigContext from "../lib/ConfigContext"

export const MailingListSignup = () => {
  const { listingsAlertUrl } = useContext(ConfigContext)

  return (
    <ActionBlock
      className={"mt-4"}
      header={<Heading priority={2}>{t("welcome.newListingEmailAlert")}</Heading>}
      background="primary-lighter"
      icon={<Icon size="3xl" symbol="mailThin" fill="transparent" />}
      actions={[
        <Link
          className="button"
          key="action-1"
          external={true}
          href={listingsAlertUrl}
          target="_blank"
        >
          {t("welcome.signUpToday")}
        </Link>,
      ]}
    />
  )
}
