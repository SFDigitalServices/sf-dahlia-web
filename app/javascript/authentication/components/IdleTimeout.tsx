import React, { useContext } from "react"

import { t } from "@bloom-housing/ui-components"

import { getHomepagePath, getSignInPath } from "../../util/routeUtil"
import UserContext from "../context/UserContext"
import BaseIdleTimeout from "./BaseIdleTimeout"

interface IdleTimeoutProps {
  onTimeout?: () => unknown
  useFormTimeout?: boolean
}

const IdleTimeout = ({ onTimeout, useFormTimeout = false }: IdleTimeoutProps) => {
  const { profile, timeOut } = useContext(UserContext)

  const handleTimeout = async (shouldTimeOut: boolean) => {
    onTimeout && (await onTimeout())
    timeOut && shouldTimeOut && timeOut()
  }

  // Only render the IdleTimeout component if the user is logged in
  if (profile && timeOut) {
    return (
      <BaseIdleTimeout
        promptTitle={t("idleTimeout.stayLoggedIn")}
        promptText={t("idleTimeout.sessionInactivityLoggedIn")}
        promptAction={t("t.continue")}
        redirectPath={getSignInPath()}
        alertMessage={t("signIn.userTokenValidationTimeout")}
        alertType={useFormTimeout ? "alert" : "notice"}
        onTimeout={() => handleTimeout(true)}
      />
    )
  }

  if (useFormTimeout) {
    return (
      <BaseIdleTimeout
        promptTitle={t("idleTimeout.stayLoggedIn")}
        promptText={t("idleTimeout.sessionInactivity")}
        promptAction={t("t.continue")}
        redirectPath={getHomepagePath()}
        alertMessage={t("idleTimeout.sessionExpired")}
        alertType={"alert"}
        onTimeout={() => handleTimeout(false)}
      />
    )
  }

  return null
}

export default IdleTimeout
