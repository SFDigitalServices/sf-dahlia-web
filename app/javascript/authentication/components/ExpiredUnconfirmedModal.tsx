import { t, Button, AppearanceStyleType } from "@bloom-housing/ui-components"
import { Dialog } from "@bloom-housing/ui-seeds"
import React, { useState } from "react"
import { EmailBanners, requestEmail } from "./EmailBanners"

export const ExpiredUnconfirmedModal = ({
  email,
  onClose,
}: {
  email: string
  onClose: () => void
}) => {
  const [emailSent, setEmailSent] = useState(false)
  const [emailSentError, setEmailSentError] = useState<string | null>(null)

  return (
    <Dialog isOpen={!!email} onClose={onClose}>
      <Dialog.Header>{t("signIn.expiredUnconfirmed.title")}</Dialog.Header>
      <EmailBanners emailSent={emailSent} emailSentError={emailSentError} />
      <Dialog.Content>{t("signIn.expiredUnconfirmed.p1")}</Dialog.Content>
      <Dialog.Content>{t("signIn.expiredUnconfirmed.p2")}</Dialog.Content>
      <Dialog.Footer>
        <Button
          type="submit"
          styleType={AppearanceStyleType.primary}
          onClick={() => requestEmail(email, setEmailSent, setEmailSentError)}
        >
          {t("signIn.expiredUnconfirmed.sendNewLink")}
        </Button>
      </Dialog.Footer>
    </Dialog>
  )
}
