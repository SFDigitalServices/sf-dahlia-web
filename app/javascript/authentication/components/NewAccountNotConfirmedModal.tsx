import { t, Button, AppearanceStyleType } from "@bloom-housing/ui-components"
import { Dialog } from "@bloom-housing/ui-seeds"
import React, { useState } from "react"
import { EmailBanners, requestEmail } from "./EmailBanners"

export const NewAccountNotConfirmedModal = ({
  email,
  onClose,
}: {
  email: string
  onClose: () => void
}) => {
  const [emailSent, setEmailSent] = useState(false)
  const [emailSentError, setEmailSentError] = useState<string | null>(null)

  return (
    <Dialog isOpen={!!email} onClose={onClose}>
      <Dialog.Header>{t("signIn.newAccount.title")}</Dialog.Header>
      <EmailBanners emailSent={emailSent} emailSentError={emailSentError} />
      <Dialog.Content>{t("signIn.newAccount.p1", { email })}</Dialog.Content>
      <Dialog.Content>{t("signIn.newAccount.p2")}</Dialog.Content>
      <Dialog.Footer>
        <Button type="submit" styleType={AppearanceStyleType.primary} onClick={onClose}>
          {t("t.ok")}
        </Button>
        <Button
          styleType={AppearanceStyleType.secondary}
          onClick={() => requestEmail(email, setEmailSent, setEmailSentError)}
        >
          {t("signIn.newAccount.sendEmailAgainButton")}
        </Button>
      </Dialog.Footer>
    </Dialog>
  )
}
