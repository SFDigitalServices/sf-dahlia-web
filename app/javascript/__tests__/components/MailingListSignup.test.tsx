import React from "react"
import renderer from "react-test-renderer"
import { MailingListSignup } from "../../components/MailingListSignup"

describe("MailingListSignup", () => {
  it("displays mailing list sign up", () => {
    const tree = renderer.create(<MailingListSignup />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
