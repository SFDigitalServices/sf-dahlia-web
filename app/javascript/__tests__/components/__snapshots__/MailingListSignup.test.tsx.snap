// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MailingListSignup displays mailing list sign up 1`] = `
<div
  className="action-block mt-4 primary-lighter action-block__block"
>
  <div
    className="action-block__head"
  >
    <div
      className="action-block__icon"
    >
      <span
        className="ui-icon ui-3xl"
        data-test-id={null}
      >
        <svg
          fill="transparent"
          height="35"
          viewBox="0 0 80 35"
          width="80"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="transparent"
            height="75"
            width="80"
          />
          <g
            clipPath="url(#clip0)"
          >
            <rect
              fill="transparent"
              height="1390"
              transform="translate(-284 -530)"
              width="1280"
            />
            <path
              d="M61.3749 24.2402L43.2061 39.7424C41.8262 40.8341 39.7563 40.8341 38.3764 39.7424L19.9775 24.2402"
              stroke="black"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeMiterlimit="10"
              strokeWidth="2"
            />
            <path
              d="M49.1858 41.271L61.375 50.878"
              stroke="black"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeMiterlimit="10"
              strokeWidth="2"
            />
            <path
              d="M32.1668 41.271L19.9775 50.878"
              stroke="black"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeMiterlimit="10"
              strokeWidth="2"
            />
            <path
              d="M14.458 24.0218V51.0961C14.458 53.9345 16.7579 56.1179 19.7477 56.1179H61.835C64.8248 56.1179 67.1247 53.9345 67.1247 51.0961V24.0218C67.1247 21.1834 64.8248 19 61.835 19H19.7477C16.7579 19 14.458 21.1834 14.458 24.0218Z"
              stroke="black"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeMiterlimit="10"
              strokeWidth="2"
            />
          </g>
          <defs>
            <clipPath
              id="clip0"
            >
              <rect
                fill="white"
                height="1390"
                transform="translate(-284 -530)"
                width="1280"
              />
            </clipPath>
          </defs>
        </svg>
      </span>
    </div>
    <div
      className="action-block__header"
    >
      <h2
        className=""
      >
        Get emailed whenever a new listing is posted.
      </h2>
    </div>
  </div>
  <div
    className="action-block__actions"
  >
    <a
      className="button"
      href=""
      target="_blank"
    >
      Sign Up today
    </a>
  </div>
</div>
`;
